synapse/daemon.py

```diff
@@ -109,32 +107,38 @@ class Genr(Share):
     async def _runShareLoop(self):
 
         # automatically begin yielding
-
         def syncloop():
-
             try:
-
-                for item in self.item:
-
+                while True:
+                    try:
+                        item = next(self.item)
+                    except StopIteration:
+                        break
+                    except Exception as e:
+                        retn = s_common.retnexc(e)
+                        mesg = ('share:data', {'share': self.iden, 'data': retn})
+                        try:
+                            s_glob.sync(self.link.tx(mesg))
+                        except Exception as e:
+                            logger.exception('Failure in communicating exception')
+                        break
                     if self.isfini:
-                        self.item.close()
                         break
 
                     retn = (True, item)
                     mesg = ('share:data', {'share': self.iden, 'data': retn})
-                    s_glob.sync(self.link.tx(mesg))
-
-            except Exception as e:
-
-                retn = s_common.retnexc(e)
-                mesg = ('share:data', {'share': self.iden, 'data': retn})
-                s_glob.sync(self.link.tx(mesg))
-
+                    try:
+                        s_glob.sync(self.link.tx(mesg))
```

Comment:
Seeing these kinda makes me wanna move toward tx() handling his exceptions and returning False so they can be a bit cleaner.